Replace --force-pull with --pull-policy

mfojtik opened the pull request:
This will replace the --force-pull with --pull-policy option. This option default to "if-not-present" which means that the Docker image will be pull only if it does not exists locally.

Valid options are "always", "never" or "if-not-present".

* `always` will always try to pull the Docker image first and then check if the local image exists
* `if-not-present` will honor the local image and only in case the local image does not exists, it will attempt to pull it
* `never` will never perform pull and always use the local image

Fixes: https://github.com/openshift/source-to-image/issues/273

mfojtik commented:
@bparees @soltysh PTAL

mfojtik commented:
[test]

soltysh commented on cmd/s2i/main.go:
```diff
@@ -203,7 +207,7 @@ $ s2i build . centos/ruby-22-centos7 hello-world-app
 	buildCmd.Flags().StringVarP(&(oldDestination), "location", "l", "", "Specify a destination location for untar operation")
 	buildCmd.Flags().StringVarP(&(cfg.Destination), "destination", "d", "", "Specify a destination location for untar operation")
-	buildCmd.Flags().BoolVar(&(cfg.ForcePull), "force-pull", true, "Always pull the builder image even if it is present locally")
+	buildCmd.Flags().VarP(&(cfg.BuilderPullPolicy), "pull-policy", "p", "Specify how to pull the builder image (always, never or if-not-present)")
```
just PullPolicy, please

mfojtik replied:
PullPolicy is the type ;-)

mfojtik replied:
maybe would be good to separate Builder and the Incremental image pulls in future

soltysh replied:
For the latter, I can agree, for the former, I don't care, we can have both names being identical. One is type and other is a field in a struct, that doesn't conflict, imho.

mfojtik replied:
@soltysh I hate having "PullPolicy PullPolicy", so will keep it this way. It is more expressive now as it describes more that this policy is for fetching the Builder image

bparees replied:
+1 to keeping this as is for both reasons @mfojtik stated.

soltysh commented on pkg/api/types.go:
```diff
@@ -230,3 +241,37 @@ const (
+// command line parameter.
+// This will also default to "if-not-present" if this parameter is not set.
+func (p *PullPolicy) String() string {
+	if len(string(*p)) == 0 {
```
Hmmm... not sure, I'd rather see when it's actually empty and not default it in some hidden place.

mfojtik replied:
@soltysh this is really just a hack to make s2i build --help show "if-not-present" as a default option, it is not used anywhere else in code

soltysh commented on pkg/build/config.go:
```diff
@@ -17,10 +17,16 @@ func GenerateConfigFromLabels(image string, config *api.Config) error {
 		source, err = d.CheckAndPullImage(image)
+	case api.PullNever:
+		source, err = d.CheckImage(image)
+	default:
```
That default should not happen. IIRC, we have a validation method (see `pkg/api/validation`) which is being called before running a build. You should create a validation there.

soltysh commented:
You're getting my LGTM, but I still dislike the string hack for printing the default value for PullPolicy :stuck_out_tongue_winking_eye:

mfojtik commented:
@soltysh it is not hack, just not ideal interface that pflags gives us :-)

mfojtik commented:
amazon hates me, [test]

soltysh commented:
Did we got banned by gh :wink: for testing too much 
[test]

bparees commented on cmd/s2i/main.go:
```diff
@@ -203,7 +207,7 @@ $ s2i build . centos/ruby-22-centos7 hello-world-app
 	buildCmd.Flags().StringVar(&(oldScriptsFlag), "scripts", "", "Specify a URL for the assemble and run scripts")
 	buildCmd.Flags().StringVarP(&(oldDestination), "location", "l", "", "Specify a destination location for untar operation")
 	buildCmd.Flags().StringVarP(&(cfg.Destination), "destination", "d", "", "Specify a destination location for untar operation")
-	buildCmd.Flags().BoolVar(&(cfg.ForcePull), "force-pull", true, "Always pull the builder image even if it is present locally")
```
deprecate, don't remove.

bparees replied:
ForcePull=true should set pull policy always, and ForcePull=false should set PullIfNotPresent.  

and we'll have to change the default, unfortunately, to ForcePull=false, otherwise it would conflict with the default of PINP for BuilderPullPolicy.

mfojtik replied:
done

bparees commented on cmd/s2i/main.go:
```diff
@@ -267,7 +275,7 @@ func newCmdRebuild(cfg *api.Config) *cobra.Command {
 	buildCmd.Flags().BoolVar(&(cfg.Incremental), "incremental", false, "Perform an incremental build")
 	buildCmd.Flags().BoolVar(&(cfg.RemovePreviousImage), "rm", false, "Remove the previous image during incremental builds")
 	buildCmd.Flags().StringVar(&(cfg.CallbackURL), "callback-url", "", "Specify a URL to invoke via HTTP POST upon build completion")
-	buildCmd.Flags().BoolVar(&(cfg.ForcePull), "force-pull", true, "Always pull the builder image even if it is present locally")
```
again deprecate

bparees commented on cmd/s2i/main.go:
```diff
@@ -324,7 +336,7 @@ func newCmdUsage(cfg *api.Config) *cobra.Command {
 	usageCmd.Flags().StringP("env", "e", "", "Specify an environment var NAME=VALUE,NAME2=VALUE2,...")
 	usageCmd.Flags().StringVarP(&(cfg.ScriptsURL), "scripts-url", "s", "", "Specify a URL for the assemble and run scripts")
 	usageCmd.Flags().StringVar(&(oldScriptsFlag), "scripts", "", "Specify a URL for the assemble and run scripts")
-	usageCmd.Flags().BoolVar(&(cfg.ForcePull), "force-pull", true, "Always pull the builder image even if it is present locally")
```
deprecate.

bparees commented on cmd/s2i/main.go:
```diff
@@ -324,7 +336,7 @@ func newCmdUsage(cfg *api.Config) *cobra.Command {
 	usageCmd.Flags().StringVarP(&(cfg.ScriptsURL), "scripts-url", "s", "", "Specify a URL for the assemble and run scripts")
 	usageCmd.Flags().StringVar(&(oldScriptsFlag), "scripts", "", "Specify a URL for the assemble and run scripts")
-	usageCmd.Flags().BoolVar(&(cfg.ForcePull), "force-pull", true, "Always pull the builder image even if it is present locally")
+	usageCmd.Flags().VarP(&(cfg.BuilderPullPolicy), "pull-policy", "p", "Specify how to pull the builder image (always, never or if-not-present)")
```
s/how/when/

(same everywhere else)

bparees commented on pkg/api/types.go:
```diff
@@ -12,6 +14,15 @@ const (
 )
 
+const (
+	// PullAlways means that we always attempts to pull the latest image.
```
attempt

bparees commented on pkg/api/types.go:
```diff
@@ -12,6 +14,15 @@ const (
+const (
+	// PullAlways means that we always attempts to pull the latest image.
+	PullAlways PullPolicy = "always"
+	// PullNever means that we never pulls an image, but only uses a local image.
```
pull
use

bparees commented on pkg/build/strategies/sti/sti.go:
```diff
@@ -323,15 +324,19 @@ func (b *STI) Exists(config *api.Config) bool {
 	}
 
-	// can only do incremental build if runtime image exists, so always pull image
-	previousImageExists, _ := b.docker.IsImageInLocalRegistry(config.Tag)
```
you've broken this behavior which was by design to pull the previous version of the application image if it didn't exist, regardless of the "force pull" flag.  

whether or not we pull the previous application image should be entirely independent of the BuilderPullPolicy.

The right way to fix this is to go ahead and introduce a PreviousImagePullPolicy.

mfojtik replied:
fixed

mfojtik commented on pkg/docker/util.go:
```diff
@@ -155,3 +159,86 @@ func parseRepositoryTag(repos string) (string, string, string) {
+// PullImage pulls the Docker image specifies by name taking the pull policy
+// into the account.
+// TODO: The 'force' option will be removed
+func PullImage(name string, d Docker, policy api.PullPolicy, force bool) (*PullResult, error) {
```
fyi this is here just to avoid writing the same switch everytime we want to pull the image (usage, incremental, build, etc)

soltysh replied:
:+1:

mfojtik commented:
@soltysh @bparees ready for the second round, summary of changes:

* Keep `--force-pull` around but default it to 'false' now (if you set it to true, you get nice deprecation warning)
* Refactored the pkg/test thing (moved docker to pkg/dockerfake_docker.go) to avoid import cycles
* Refactored image pulling to avoid code duplication

mfojtik commented:
As a follow up we should:

* Refactor the pkg/test entirely (ideally get rid of that package as it causes import cycles)
* Reorganize the packages a little bit (currently there are things mixed together like CheckAllowedUsers in pkg/docker where it should really be in util/user, etc... there are much examples like this and that require more general refactoring)

soltysh commented on cmd/s2i/main.go:
```diff
@@ -122,6 +122,10 @@ $ s2i build . centos/ruby-22-centos7 hello-world-app
 			}
 
+			if cfg.ForcePull == true {
+				glog.Warning("DEPRECATED: The '--force-pull' option is deprecated. Use '--pull-policy' instead")
```
Make all the deprecation info identical, see [here](https://github.com/openshift/source-to-image/blob/master/cmd/s2i/main.go#L163) and [here](https://github.com/openshift/source-to-image/blob/master/cmd/s2i/main.go#L167). Doing the latter two your way is also a good option, or making a function for that would be the best :wink:

soltysh commented on cmd/s2i/main.go:
```diff
@@ -122,6 +122,10 @@ $ s2i build . centos/ruby-22-centos7 hello-world-app
 				}
 			}
 
+			if cfg.ForcePull == true {
```
`if cfg.ForcePull {` should be enough.

soltysh commented on cmd/s2i/main.go:
```diff
@@ -267,7 +276,8 @@ func newCmdRebuild(cfg *api.Config) *cobra.Command {
 	buildCmd.Flags().BoolVar(&(cfg.RemovePreviousImage), "rm", false, "Remove the previous image during incremental builds")
 	buildCmd.Flags().StringVar(&(cfg.CallbackURL), "callback-url", "", "Specify a URL to invoke via HTTP POST upon build completion")
-	buildCmd.Flags().BoolVar(&(cfg.ForcePull), "force-pull", true, "Always pull the builder image even if it is present locally")
+	buildCmd.Flags().BoolVar(&(cfg.ForcePull), "force-pull", false, "DEPRECATED: Always pull the builder image even if it is present locally")
```
Again, it would be good to make all the detracted flags look identical, preferring your version.

mfojtik replied:
done

soltysh commented on pkg/api/types.go:
```diff
@@ -12,6 +14,24 @@ const (
+
+	// DefaultPreviousImagePullPolicy specifies policy for pulling the previously
+	// build Docker image when doing incremental build
+	DefaultPreviousImagePullPolicy = PullAlways
```
If you're providing that option, how about making it accessible for users to set?

mfojtik replied:
1) how would you call it? --previous-pull-policy? ;-)
2) none if of the names I have in my mind are good
3) you don't want to confuse people (at least for now)

mfojtik replied:
@soltysh so if you

1) come up with a use-case where setting this is actually useful
2) and give me reasonable option name

I can expose it to users :-)

soltysh replied:
Yeah, the usual naming problem :wink:

soltysh commented on pkg/api/types.go:
```diff
@@ -65,6 +85,19 @@ type Config struct {
+	// This is now deprecated by BuilderPullPolicy and will be removed soon.
+	// Setting this to 'true' equals setting BuilderPullPolicy to 'PullAlways'.
+	// Setting this to 'false' equals setting BuilderPullPolicy to 'PullIfNotPresent'
+	ForcePull bool
```
Add a deprecation comment.

mfojtik replied:
This is now deprecated by BuilderPullPolicy and will be removed soon. 

this is not enough?

soltysh replied:
I'm blind, sorry.

soltysh commented on pkg/build/strategies/sti/sti.go:
```diff
@@ -323,15 +323,18 @@ func (b *STI) Exists(config *api.Config) bool {
+	policy := config.PreviousImagePullPolicy
+	if len(policy) == 0 {
+		policy = api.DefaultPreviousImagePullPolicy
+	}
```
Shouldn't that defaulting happen in `cmd/s2i/main.go` like all the others? I told you to create defaults :stuck_out_tongue_winking_eye:

mfojtik replied:
what about origin? i think we want to set default there as well, although in origin it will be PullAlways

soltysh replied:
It can always have own defaulting function, remember origin is using slightly different flow, and having such defaulting function gives us confidence when and what happens.

mfojtik replied:
@soltysh I think defaulter will be also on my list for the future refactoring :-) but currently we have just PullPolicy we have to default and don't want to spend a lot of time refactoring code just to default one option ;-)

soltysh replied:
Create issues, so we don't forget.

soltysh commented:
Nits mostly, other than that LGTM.

bparees commented on pkg/api/describe/describer.go:
```diff
@@ -39,7 +39,7 @@ func DescribeConfig(config *api.Config) string {
 		}
 		fmt.Fprintf(out, "Remove Old Build:\t%s\n", printBool(config.RemovePreviousImage))
-		fmt.Fprintf(out, "Force Pull:\t%s\n", printBool(config.ForcePull))
+		fmt.Fprintf(out, "Pull Policy:\t%s\n", config.BuilderPullPolicy)
```
Builder Pull Policy

bparees commented:
needs squash

mfojtik commented:
@bparees i want to remove the last commit before push, it is there to help debug the integration failure which is not clear to me...

soltysh commented on pkg/docker/util.go:
```diff
@@ -155,3 +159,85 @@ func parseRepositoryTag(repos string) (string, string, string) {
+		image, err = d.CheckAndPullImage(name)
+	case api.PullAlways:
+		image, err = d.PullImage(name)
+		fallthrough
```
:8ball:

soltysh commented on cmd/s2i/main.go:
```diff
@@ -200,10 +211,11 @@ $ s2i build . centos/ruby-22-centos7 hello-world-app
 	buildCmd.Flags().StringVarP(&(cfg.ScriptsURL), "scripts-url", "s", "", "Specify a URL for the assemble and run scripts")
-	buildCmd.Flags().StringVar(&(oldScriptsFlag), "scripts", "", "Specify a URL for the assemble and run scripts")
+	buildCmd.Flags().StringVar(&(oldScriptsFlag), "scripts", "", "DEPRECATED: Specify a URL for the assemble and run scripts")
 	buildCmd.Flags().StringVarP(&(oldDestination), "location", "l", "", "Specify a destination location for untar operation")
```
Location is also deprecated, pretty please add DEPRECATED in front of it.

mfojtik replied:
i'm doing that in follow up with main.go refactoring.

soltysh replied:
k

soltysh commented on hack/test-stirunimage.sh:
```diff
@@ -34,16 +34,13 @@ fi
-
-sti build git://github.com/bparees/openshift-jee-sample openshift/wildfly-8-centos test-jee-app --run=true &> "${STI_ROOT}/hack/sti-run.log" &
+echo "Running 'sti build --run=true ...'"
+s2i build git://github.com/bparees/openshift-jee-sample openshift/wildfly-8-centos test-jee-app --run=true &> "${STI_ROOT}/hack/sti-run.log" &
```
This should be also fixed to use the `openshift/wildfly-81-centos7` instead, can be part of that refactor of yours.

mfojtik replied:
@soltysh this is a tribute to @bparees :-)

soltysh replied:
I know, I'm just saying to use newer image he still maintains :wink:

bparees replied:
yeah please change it :)

bparees replied:
i'd even consider changing it to the php app or something because that'll build faster than the JEE app does, so tests will be faster.

mfojtik replied:
i'll do that as follow up, don't want to make this PR fatter :-)

soltysh commented:
LGTM, @bparees merge it :smile:

bparees commented on pkg/build/strategies/sti/sti.go:
```diff
@@ -401,7 +404,7 @@ func (b *STI) Execute(command string, config *api.Config) error {
 		Stdout:          outWriter,
 		Stderr:          errWriter,
-		PullImage:       config.ForcePull,
+		PullImage:       config.BuilderPullPolicy == api.PullAlways || config.BuilderPullPolicy == api.PullIfNotPresent,
```
doesn't this always result in a Pull for PullIfNotPresent?  In the case of PullIfNotPresent, don't you need to manually check if the image exists locally first, and then set this to true or false accordingly?

bparees replied:
alternatively if you assume by this point in the code you've already called PullImage (which will have pulled the image if appropriate), then you should always be passing "false" here because the right image (freshly pulled or not) will already exist locally.

mfojtik replied:
right, false should be safe because the image should be pulled in all cases

bparees commented on pkg/docker/util.go:
```diff
@@ -155,3 +159,85 @@ func parseRepositoryTag(repos string) (string, string, string) {
+
+	// When pull policy not specified, default to DefaultPullPolicy
+	if len(policy) == 0 {
+		policy = api.DefaultBuilderPullPolicy
```
you're defaulting to the Builder policy, but this method might be invoked for Builder images or Previous images...  that's a problem.  You should just require a policy be passed in and error out otherwise.

mfojtik replied:
@bparees the other place this is used is to pull previous image that has the policy explicitly set.
I think we should not default to any policy and just return error when the policy is not set.

soltysh replied:
I'd again suggest creating defaults, but I remember you've it already in the followup PR :wink:

mfojtik replied:
It is not needed as we run PullImage() before so the image should always be
available or this should error out.

On Fri, Nov 6, 2015 at 2:14 PM, Maciej Szulik <notifications@github.com>
wrote:

> In pkg/docker/util.go
> <https://github.com/openshift/source-to-image/pull/324#discussion_r44134559>
> :
>
> > @@ -155,3 +159,85 @@ func parseRepositoryTag(repos string) (string, string, string) {
> >  	}
> >  	return repos, "", ""
> >  }
> > +
> > +// PullImage pulls the Docker image specifies by name taking the pull policy
> > +// into the account.
> > +// TODO: The 'force' option will be removed
> > +func PullImage(name string, d Docker, policy api.PullPolicy, force bool) (*PullResult, error) {
> > +	// TODO: Remove this after we deprecate --force-pull
> > +	if force {
> > +		policy = api.PullAlways
> > +	}
> > +
> > +	// When pull policy not specified, default to DefaultPullPolicy
> > +	if len(policy) == 0 {
> > +		policy = api.DefaultBuilderPullPolicy
>
> I'd again suggest creating defaults, but I remember you've it already in
> the followup PR [image: :wink:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/source-to-image/pull/324/files#r44134559>.
>



-- 
-----------------------------------------------
Michal Fojtik <mi@mifo.sk <mfojtik@mifo.sk>>
Red Hat OpenShift, Engineering

bparees replied:
> @bparees the other place this is used is to pull previous image that has the policy explicitly set.
I think we should not default to any policy and just return error when the policy is not set.

isn't that what i suggested? :)

mfojtik replied:
Yes, it is :-)

On Fri, Nov 6, 2015 at 4:00 PM, Ben Parees <notifications@github.com> wrote:

> In pkg/docker/util.go
> <https://github.com/openshift/source-to-image/pull/324#discussion_r44144586>
> :
>
> > @@ -155,3 +159,85 @@ func parseRepositoryTag(repos string) (string, string, string) {
> >  	}
> >  	return repos, "", ""
> >  }
> > +
> > +// PullImage pulls the Docker image specifies by name taking the pull policy
> > +// into the account.
> > +// TODO: The 'force' option will be removed
> > +func PullImage(name string, d Docker, policy api.PullPolicy, force bool) (*PullResult, error) {
> > +	// TODO: Remove this after we deprecate --force-pull
> > +	if force {
> > +		policy = api.PullAlways
> > +	}
> > +
> > +	// When pull policy not specified, default to DefaultPullPolicy
> > +	if len(policy) == 0 {
> > +		policy = api.DefaultBuilderPullPolicy
>
> @bparees <https://github.com/bparees> the other place this is used is to
> pull previous image that has the policy explicitly set.
> I think we should not default to any policy and just return error when the
> policy is not set.
>
> isn't that what i suggested? :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/source-to-image/pull/324/files#r44144586>.
>



-- 
-----------------------------------------------
Michal Fojtik <mi@mifo.sk <mfojtik@mifo.sk>>
Red Hat OpenShift, Engineering

bparees commented:
lgtm.

@gabemontero heads up, this may cause some conflicts in your branch.  hopefully only around the test scripts.

[merge]

soltysh commented:
:tada:
